### PR TITLE
Data sharing part 1

### PR DESCRIPTION
**Pull request by mrobertshaw**
Bulk of the generic data sharing logic including enhancements to the security service to break top down security model and allow for sharing via object scope access 1-many process

### CODE REVIEW

**Review by PeterSpainOXIL — APPROVED** [review]

File: php/src/Services/Security/SecurityService.php [hunk 1]
```diff
@@ -383,30 +388,93 @@ public function checkLoggedInObjectAccess($object, $accessMode = self::ACCESS_RE
...
-                    break;
+
+
             }
 
         }
 
-        return $access;
+        // Return at least one true
+        return sizeof($accessGroupGranted) == 0 || in_array(true, $accessGroupGranted);
```
Comment (peteroxil) [inline php/src/Services/Security/SecurityService.php:441]: ! Why is this boolean or? If accessGroupGranted is empty why does the user have permission?

File: php/src/Traits/Security/Sharable.php [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+     * @return string[]
+     */
+    public function returnValidObjectScopeAccesses($recipientScope = null) {
+        $scopedObjects = $recipientScope ? ObjectArrayUtils::filterArrayOfObjectsByMember("recipientScope", $this->objectScopeAccesses ?? [], $recipientScope) :
+            $this->objectScopeAccesses;
+        return array_filter($scopedObjects, function ($object) {
+            return !$object->getExpiryDate() || $object->getExpiryDate() >= new \DateTime();
+        });
+    }
+
```
Comment (peteroxil) [inline php/src/Traits/Security/Sharable.php:56]:
```php
ObjectArrayUtils::filterArrayOfObjectsByMember("recipientScope", $this->objectScopeAccesses ?? [], $recipientScope);
```

vs the more explicit

```php
array_filter(
    $this->objectScopeAccesses ?? [],
    fn($access) => $recipientType == $access->getRecipientScope());
```

File: php/src/ValueObjects/Security/SharableItem.php [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+    /**
+     * @var string
+     */
+    private $sharerLogo;
+
+
+    public function __construct(mixed $sharable) {
+        if (!in_array(Sharable::class, class_uses(get_class($sharable)))) {
+            throw new WrongParameterTypeException("Sharable Items must be constructed with sharable objects");
+        }
```
Comment (peteroxil) [inline php/src/ValueObjects/Security/SharableItem.php:38]: `class_uses` doesn't return traits of superclasses. We may need recursive function here.

**Review by peteroxil — COMMENTED** [review]
Looks good to me. I've left some comments asking for clarity. ScopeAccessGroup specifically is a little mysterious - I assume this maps to a frontend thing which is a group of Accounts/Projects that you give permissions to?

File: php/src/Objects/Security/ObjectScopeAccess.php
[the remark this answers is not in the document]
Comment (mrobertshaw) [reply php/src/Objects/Security/ObjectScopeAccess.php:1]: I'm sure write and grant could be stricter booleans.  Null does indeed mean the same as false here 

The access group is actually a relational hash to group related scope access items together

File: php/src/Services/Security/ActiveRecordInterceptor.php [hunk 4]
```diff
@@ -149,6 +159,32 @@ public function executeInsecure($callable) {
...
+        $previousWhiteListed = $this->whitelistedReadAccounts;
+
+        $this->whitelistedReadAccounts[] = $accountId;
+
+        // Run the callable
+        try {
+            $result = $callable();
+            $this->whitelistedReadAccounts = $previousWhiteListed;
+            return $result;
+        } catch (\Throwable $e) {
```
[the remark this answers is not in the document]
Comment (mrobertshaw) [reply php/src/Services/Security/ActiveRecordInterceptor.php:180]: Yes, could be backed off to Exception

File: php/src/Services/Security/ObjectScopeAccessService.php [hunk 5]
```diff
@@ -0,0 +1,347 @@
...
+    public function assignScopeAccessGroupsToObject(string $objectClassName, string $objectPrimaryKey, array $scopeAccessGroups) {
+
+        // Check that the class being assigned is sharable
+        $this->checkObjectSharableByLoggedInUser($objectClassName, $objectPrimaryKey);
+
+        // Save all object scope accesses
+        foreach ($scopeAccessGroups as $group) {
+            foreach ($group->getScopeAccesses() as $objectScopeItem) {
+                $objectScopeAccess = new ObjectScopeAccess($objectScopeItem->getScope(), $objectScopeItem->getItemIdentifier(), $group->getGroupName(), $group->getWriteAccess(), $group->getGrantAccess(), $group->getExpiryDate(), $objectClassName, $objectPrimaryKey);
+                $objectScopeAccess->save();
```
[the remark this answers is not in the document]
Comment (mrobertshaw) [reply php/src/Services/Security/ObjectScopeAccessService.php:89]: Hopefully this is an edge case but yes you can overwrite in retry

File: php/src/Services/Security/ObjectScopeAccessService.php [hunk 6]
```diff
@@ -0,0 +1,347 @@
...
+     *
+     * @objectInterceptorDisabled
+     */
+    public function listInvitationsForSharedObject(string $objectClassName, string $objectPrimaryKey) {
+
+        // Check the class being assigned is sharable
+        $this->checkObjectSharableByLoggedInUser($objectClassName, $objectPrimaryKey);
+
+        // Grab invitations
+        $invitations = $this->pendingActionService->getAllPendingActionsForTypeAndObjectId("OBJECT_SHARING_INVITE", $objectPrimaryKey, $objectClassName);
```
[the remark this answers is not in the document]
Comment (mrobertshaw) [reply php/src/Services/Security/ObjectScopeAccessService.php:274]: I went back and forth on this but as the pending actions are fairly loose and not from a fixed set it's probably not so useful to have a constant / enumerated type

File: php/src/Services/Security/SecurityService.php (hunk 1 above)
Comment (mrobertshaw) [reply php/src/Services/Security/SecurityService.php:441]: This is to support previous behaviour where there are no access groups in play.  I suspect if it gets this far it won't have been vetoed

File: php/src/ValueObjects/Security/SharableItem.php (hunk 3 above)
Comment (mrobertshaw) [reply php/src/ValueObjects/Security/SharableItem.php:38]: That's a good point and probably bears sorting - I think I wrote a recursive function for this

**Pull request merged**